### PR TITLE
David update gce instance script

### PR DESCRIPTION
**Pull request by davidzhu22**
David update gce instance script

### CODE REVIEW

File: hack/test-gce/fornaxcore_install.sh [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+echo -e "## DISABLING FIREWALL\n"
+sudo ufw disable
+sudo swapoff -a
+
+basic_install() {
+    echo -e "## INSTALL BASIC TOOL"
+    sudo apt-get -y update
+    sudo apt -y install build-essential
+    sudo apt -y install curl
+    sudo apt-get -y install vim
```
Comment (guowenjian90) [inline hack/test-gce/fornaxcore_install.sh:20]: why you have to install these tools.

File: hack/test-gce/fornaxcore_install.sh [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    echo -e "## DONE BASIC TOOL\n"
+}
+
+
+docker_install(){
```
Comment (guowenjian90) [inline hack/test-gce/fornaxcore_install.sh:25]: fornax core do not need docker

File: hack/test-gce/fornaxcore_install.sh [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	mkdir -p bin src pkg
+	cd src
+	mkdir -p centaurusinfra.io
+	cd centaurusinfra.io
+	# pushd $HOME/go/src/centaurusinfra.io
+    sudo git clone https://github.com/CentaurusInfra/fornax-serverless.git
+    # pushd $HOME/go/src/centaurusinfra.io/fornax-serverless
+	cd fornax-serverless
+	sudo chown -R $USER: .
+    make all
```
Comment (guowenjian90) [inline hack/test-gce/fornaxcore_install.sh:87]: I thought we agreed do not need get source code and compile it locally

File: hack/test-gce/gce_create.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+read account_number
+echo -e "\n"
+
+
+# davidzhu-fornaxcore
+inst=`gcloud compute instances list --project quark-serverless --format="table(name)" --filter="name=davidzhu-fornaxcore" | awk '{print $1}'`
+if [[ $inst == "" ]];
+then
+	echo -e "will create a instance: davidzhu-fornaxcore"
+    gcloud compute instances create davidzhu-fornaxcore --zone=us-central1-a --machine-type=e2-medium --network-interface=network-tier=PREMIUM,subnet=default --maintenance-policy=MIGRATE --provisioning-model=STANDARD --service-account=$account_number-compute@developer.gserviceaccount.com --scopes=https://www.googleapis.com/auth/cloud-platform --tags=http-server,https-server --create-disk=auto-delete=yes,boot=yes,device-name=davidzhu-instance-1,image=projects/ubuntu-os-cloud/global/images/ubuntu-2004-focal-v20220927,mode=rw,size=50,type=projects/quark-serverless/zones/us-central1-a/diskTypes/pd-ssd --no-shielded-secure-boot --shielded-vtpm --shielded-integrity-monitoring --reservation-affinity=any
```
Comment (guowenjian90) [inline hack/test-gce/gce_create.sh:20]: avoid using personal name in an instance name, again, please make machine- type a variable, 
BTW, how big is e2-medium? fornaxcore require more cpu and memory than node agent.

File: hack/test-gce/gce_create.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+fi
+
+
+# using for loop to create node agentinstance, for example: davidzhu-nodeagent1, 2, 3...
+# instance_num=2
+for ((i = 1; i<=$instance_num; i++))
+do
+    instance_name='davidzhu-nodeagent-'$i
+    echo -e "created $instance_name \n"
+    gcloud compute instances create $instance_name --zone=us-central1-a --machine-type=e2-medium --network-interface=network-tier=PREMIUM,subnet=default --maintenance-policy=MIGRATE --provisioning-model=STANDARD --service-account=$account_number-compute@developer.gserviceaccount.com --scopes=https://www.googleapis.com/auth/cloud-platform --tags=http-server,https-server --create-disk=auto-delete=yes,boot=yes,device-name=davidzhu-instance-1,image=projects/ubuntu-os-cloud/global/images/ubuntu-2004-focal-v20220927,mode=rw,size=50,type=projects/quark-serverless/zones/us-central1-a/diskTypes/pd-ssd --no-shielded-secure-boot --shielded-vtpm --shielded-integrity-monitoring --reservation-affinity=any
```
Comment (guowenjian90) [inline hack/test-gce/gce_create.sh:32]: same as above,

File: hack/test-gce/gce_delete.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    read instance_num
+    echo -e "\n"
+
+    # delete virtual machine instance
+    gcloud compute instances delete davidzhu-fornaxcore --zone=us-central1-a
+
+    # instance_num=2
+    for ((i = 1; i<=$instance_num; i++))
+    do
+        instance_name='davidzhu-nodeagent-'$i
```
Comment (guowenjian90) [inline hack/test-gce/gce_delete.sh:18]: same as above

File: hack/test-gce/nodeagent_install.sh [hunk 7]
```diff
@@ -0,0 +1,282 @@
+#! /bin/bash
+
+set -e
+
+#Golang version
+GO_VERSION=${GO_VERSION:-"1.18.7"}
+
+pushd $HOME
+echo -e "## SETTING UP THE HOSTNAME NODEAGENT-A\n"
+sudo hostnamectl set-hostname nodeagent-a
```
Comment (guowenjian90) [inline hack/test-gce/nodeagent_install.sh:10]: why do we set host name, and it set same name for all node agents?

File: hack/test-gce/nodeagent_install.sh [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+echo -e "## DISABLING FIREWALL\n"
+sudo ufw disable
+sudo swapoff -a
+
+basic_install() {
+    echo -e "## INSTALL BASIC TOOL"
+    sudo apt-get -y update
+    sudo apt -y install build-essential
+    sudo apt -y install curl
+    sudo apt-get -y install vim
```
Comment (guowenjian90) [inline hack/test-gce/nodeagent_install.sh:20]: as above, why we need them?

File: hack/test-gce/nodeagent_install.sh [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        },
+        "quark_d": {
+            "path": "/usr/local/bin/quark_d"
+        },
+        "runsc": {
+            "path": "/usr/local/bin/runsc"
+        },
+        "kata-runtime": {
+            "path": "/usr/bin/kata-runtime"
+        }
```
Comment (guowenjian90) [inline hack/test-gce/nodeagent_install.sh:74]: we only need runc.

File: hack/test-gce/nodeagent_install.sh [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+	  mkdir -p bin src pkg
+	  cd src
+	  mkdir -p centaurusinfra.io
+	  cd centaurusinfra.io
+	  # pushd $HOME/go/src/centaurusinfra.io
+    sudo git clone https://github.com/CentaurusInfra/fornax-serverless.git
+    # pushd $HOME/go/src/centaurusinfra.io/fornax-serverless
+	  cd fornax-serverless
+	  sudo chown -R $USER: .
+    make all
```
Comment (guowenjian90) [inline hack/test-gce/nodeagent_install.sh:246]: same as above, do not need to build code in each node agent.

File: hack/test-gce/gce_create.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	echo -e "will create a instance: davidzhu-fornaxcore"
+    gcloud compute instances create davidzhu-fornaxcore --zone=us-central1-a --machine-type=e2-medium --network-interface=network-tier=PREMIUM,subnet=default --maintenance-policy=MIGRATE --provisioning-model=STANDARD --service-account=$account_number-compute@developer.gserviceaccount.com --scopes=https://www.googleapis.com/auth/cloud-platform --tags=http-server,https-server --create-disk=auto-delete=yes,boot=yes,device-name=davidzhu-instance-1,image=projects/ubuntu-os-cloud/global/images/ubuntu-2004-focal-v20220927,mode=rw,size=50,type=projects/quark-serverless/zones/us-central1-a/diskTypes/pd-ssd --no-shielded-secure-boot --shielded-vtpm --shielded-integrity-monitoring --reservation-affinity=any
+else
+	echo -e "instance: $inst already exist"
+fi
+
+
+# using for loop to create node agentinstance, for example: davidzhu-nodeagent1, 2, 3...
+# instance_num=2
+for ((i = 1; i<=$instance_num; i++))
```
Comment (guowenjian90) [inline hack/test-gce/gce_create.sh:28]: this has to be executed in parallel, other wise, if one host took 5 minutes to finish, 100 hosts take 500 minutes, nearly 8 hours to setup, it's too slow, https://cloud.google.com/compute/docs/instances/startup-scripts/linux

**Review by guowenjian90 — COMMENTED** [review]
check my comments, there are multiple things have to be addressed.

File: hack/test-gce/fornaxcore_install.sh (hunk 1 above)
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:20]: Brand new vm or instance need these basic tool

File: hack/test-gce/fornaxcore_install.sh (hunk 3 above)
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:87]: This is for our test verion for internal test. I will create a deployed version, no source code build

File: hack/test-gce/gce_create.sh (hunk 4 above)
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:20]: This is for internal test version. easy find instance in GCE dashboard in my interface.
For e2-medum. for test purpose for now. this script main test flow, not actually vm in the production env.

File: hack/test-gce/nodeagent_install.sh (hunk 7 above)
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh:10]: I comment these two line

File: hack/test-gce/nodeagent_install.sh (hunk 9 above)
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh:74]: this is our basic setting for test environment setting. not any affect test. default is runc

File: hack/test-gce/nodeagent_install.sh (hunk 8 above)
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh:20]: Brand new vm or instance need these basic tool

File: hack/test-gce/nodeagent_install.sh (hunk 10 above)
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh:246]: This is for our test verion for internal test. I will create a deployed version, no source code build

File: hack/test-gce/gce_delete.sh (hunk 6 above)
Comment (davidzhu22) [reply hack/test-gce/gce_delete.sh:18]: I will use async to do it in deployed sh file

File: hack/test-gce/gce_create.sh (hunk 11 above)
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:28]: I will look into the link which you provide. I will do async in the future and improve performance

File: hack/test-gce/fornaxcore_install.sh (hunk 2 above)
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:25]: I remove docker install function

File: hack/test-gce/gce_create.sh (hunk 5 above)
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:32]: This is for internal test version. easy find instance in GCE dashboard in my interface.
For e2-medum. for test purpose for now. this script main test flow, not actually vm in the production env.

**Review by guowenjian90 — COMMENTED** [review]
general comments,
1/ remove not used code
2/ please create a README to help understand how to use these scripts. 
I think our purpose is one entry point script will finish all installation.
3/ do not use personal name as resource name in project

File: hack/test-gce/fornaxcore_deploy.sh [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    then
+       echo -e "## ETCD IS ALREADY INSTALLED\n"
+    else
+       echo -e "##INSTALLING ETCD"
+       sudo apt-get -y install etcd > /dev/null 2>&1
+       echo -e "## ETCD INSTALLED\n"
+   fi
+}
+
+golang_tools(){
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_deploy.sh:37]: Yes, I will do right now

[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_deploy.sh:37]: done

File: hack/test-gce/fornaxcore_deploy.sh [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+basic_install
+
+etcd_install
+
+kubectl_install
+
+# golang_tools
+
+# fornaxcore_build
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_deploy.sh:113]: solved

File: hack/test-gce/fornaxcore_install.sh [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+#Golang version
+GO_VERSION=${GO_VERSION:-"1.18.7"}
+
+pushd $HOME
+# echo -e "## SETTING UP THE HOSTNAME FORNAXCORE-A\n"
+# sudo hostnamectl set-hostname fornaxcore-a
+echo -e "## DISABLING FIREWALL\n"
+sudo ufw disable
+sudo swapoff -a
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:13]: swapoff disables swapping on the specified devices and files. When the -a flag is given, swapping is disabled on all known swap devices and files. we can easy to copy file and run at bash file

File: hack/test-gce/gce_create.sh [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+do
+    instance_name='davidzhu-nodeagent-'$i
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:30]: Yes, I will correct it

File: hack/test-gce/gce_create.sh [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+read account_number
+echo -e "\n"
+
+
+# davidzhu-fornaxcore
+inst=`gcloud compute instances list --project quark-serverless --format="table(name)" --filter="name=davidzhu-fornaxcore" | awk '{print $1}'`
+if [[ $inst == "" ]];
+then
+	echo -e "will create a instance: davidzhu-fornaxcore"
+    gcloud compute instances create davidzhu-fornaxcore --project=quark-serverless --zone=us-central1-a --machine-type=e2-medium --network-interface=network-tier=PREMIUM,subnet=default --maintenance-policy=MIGRATE --provisioning-model=STANDARD --service-account=$account_number-compute@developer.gserviceaccount.com --scopes=https://www.googleapis.com/auth/cloud-platform --tags=http-server,https-server --create-disk=auto-delete=yes,boot=yes,device-name=davidzhu-instance-1,image=projects/ubuntu-os-cloud/global/images/ubuntu-2004-focal-v20220927,mode=rw,size=50,type=projects/quark-serverless/zones/us-central1-a/diskTypes/pd-ssd --no-shielded-secure-boot --shielded-vtpm --shielded-integrity-monitoring --reservation-affinity=any &
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:20]: reduce test input when start test environment. this test for internal use

File: hack/test-gce/gce_create_deploy.sh [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+read instance_num
+echo -e "\n"
+
+echo -e "## Enter account number which you want to created VM in your test:"
+read account_number
+echo -e "\n"
+
+instance_create() {
+    # davidzhu-fornaxcore
+    inst=`gcloud compute instances list --project quark-serverless --format="table(name)" --filter="name=davidzhu-fornaxcore" | awk '{print $1}'`
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_create_deploy.sh:16]: I remove all personal name

File: hack/test-gce/fornaxcore_deploy.sh [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        echo -e "## kubectl classic IS ALREADY INSTALLED\n"
+     else
+        echo -e "##INSTALLING kubectl"
+        sudo snap install kubectl --classic > /dev/null 2>&1
+        echo -e "## kubectl INSTALLED\n"
+    fi
+}
+
+
+fornaxcore_build(){
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_deploy.sh:72]: I will do it now

[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_deploy.sh:72]: removed

File: hack/test-gce/fornaxcore_install.sh [hunk 19]
```diff
@@ -0,0 +1,103 @@
+#! /bin/bash
+
+set -e
+
+#Golang version
+GO_VERSION=${GO_VERSION:-"1.18.7"}
+
+pushd $HOME
+# echo -e "## SETTING UP THE HOSTNAME FORNAXCORE-A\n"
+# sudo hostnamectl set-hostname fornaxcore-a
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:10]: remove all which have these two line

File: hack/test-gce/gce_create.sh (hunk 15 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_create.sh:30]: remove personal name in all file

File: hack/test-gce/fornaxcore_install.sh [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+    then
+       echo -e "## ETCD IS ALREADY INSTALLED\n"
+    else
+       echo -e "##INSTALLING ETCD"
+       sudo apt-get -y install etcd > /dev/null 2>&1
+       echo -e "## ETCD INSTALLED\n"
+   fi
+}
+
+golang_tools(){
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:37]: solved

File: hack/test-gce/fornaxcore_install.sh [hunk 21]
```diff
@@ -0,0 +1,101 @@
+#! /bin/bash
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh:1]: this is only use for build fornaxcore server with compile the code. not for deployed. when we setup builder server use it.

File: hack/test-gce/nodeagent_install.sh [hunk 22]
```diff
@@ -0,0 +1,278 @@
+#! /bin/bash
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh:1]: this is only use for build nodeagent server with compile the code. not for deployed. when we setup builder server use it.

File: hack/test-gce/gce_run.md [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+1. login fornaxcore machine and verify fornaxcore service is started. After login, run the following script and see fornaxcore service if it is started.
+
+```script
+ps -aux | grep fornaxcore
+```
+
+2. pickup one of nodeagent, login nodeagent machine and verify nodeagent service is started. After login, run the following script and see nodeagent serviceif it is started.
+
+```script
+ps -aus | grep nodeagent
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md:56]: let me think about how we verify all instance is running and no need to login in each instance.

File: hack/test-gce/gce_run.md [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+If you already have machine and with source file ready, and have all build execut file on  ~/go/src/CentaurusInfra/fornax-serverlessan/bin,
+you can copy following file to $HOME(or create new file with the same name and copy following file content to the file). Or go to /hach/gce_test/
+
+```script
+gce_create_deploy.sh
+fornaxcore_deploy.sh
+nodeagent_deploy.sh
+fornaxcore_start.sh
+nodeagent_start.sh
+serverless_start.sh
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md:17]: I will reorganize this part. may load all source code the server and run at /hack/gce-test directory

File: hack/test-gce/gce_run.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
+# Serverless gce test environment platform setup and running steps
+
+## Overview
+This doc represent how to setup gce serverless test envrironment and detail steps.
+
+
+## Machine and file Prepare
+If you already have machine and with source file ready, and have all build execut file on  ~/go/src/CentaurusInfra/fornax-serverlessan/bin,
+you can copy following file to $HOME(or create new file with the same name and copy following file content to the file). Or go to /hach/gce_test/
```
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md:9]: I will reorganize this part. may load all source code the server and run at /hack/gce-test directory

File: hack/test-gce/fornaxcore_install.sh (hunk 21 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/fornaxcore_install.sh]: move to hack directory

File: hack/test-gce/gce_run.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md]: rewrite this part

File: hack/test-gce/gce_run.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md]: add another script code to solved this issue

File: hack/test-gce/nodeagent_install.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/nodeagent_install.sh]: move to another folder : hach

File: hack/test-gce/gce_run.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (davidzhu22) [reply hack/test-gce/gce_run.md]: solved this issues

File: hack/test-gce/gce_func.sh [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+            scp -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -r $HOME/go/src/centaurusinfra.io/fornax-serverless/bin/nodeagent  ubuntu@$name:$HOME/go/src/centaurusinfra.io/fornax-serverless/bin/
+        fi
+    done   
+}
+
+# delete all app and clear node in the instance
+delete_app() {
+    for i in {0..100}
+    do
+        kubectl delete application --kubeconfig kubeconfig --namespace fornaxtest echoserver$i &
```
Comment (guowenjian90) [inline hack/test-gce/gce_func.sh:77]: btw, new fornaxtest use shorter name for application, it's just 'echo', not echoserver anymore, you can make correspondent change in all kubectl commands.

**Review by guowenjian90 — APPROVED** [review]
LGTM

**Comment by davidzhu22** [comment]
leader approval the code and merge in code

**Pull request merged**